meson.build

```diff
@@ -0,0 +1,216 @@
+project(
+    'cloud-init',
+    version: run_command('./tools/read-version', check: true).stdout().strip(),
+    meson_version: '>=0.63.0',
+    default_options: [
+      # The default can yield broken results.
+      'python.install_env=auto'
+    ]
+)
+completions = dependency('bash-completion')
+systemd = dependency('systemd')
+system = host_machine.system()
+
+libdir = get_option('libdir')
+pkgdatadir = get_option('datadir')
+sysconfdir = get_option('sysconfdir')
+INIT_SYSTEM = get_option('init_system')
+VALID_INIT_SYSTEMS = [
+  'systemd', 'sysvinit', 'sysvinit_deb', 'sysvinit_dragonfly',
+  'sysvinit_freebsd', 'sysvinit_netbsd', 'sysvinit_openbsd'
+]
+if not VALID_INIT_SYSTEMS.contains(INIT_SYSTEM)
+  error(
+    'Invalid init_system supplied: "@0@". Must be one of: @1@'.format(
+      INIT_SYSTEM, VALID_INIT_SYSTEMS)
+  )
+endif
+
+
+DISTRO = get_option('distro')
+BSD_DISTROS = ['dragonfly', 'freebsd', 'netbsd', 'openbsd']
+RHEL_DISTROS = [
+  'almalinux', 'centos', 'cloudlinux', 'eurolinux', 'fedora', 'miraclelinux',
+  'rhel', 'rocky', 'virtuozzo'
+]
+VALID_DISTROS = [
+  'alpine', 'amazon', 'debian', 'mariner', 'opencloudos', 'openeuler',
+  'photon', 'ubuntu', 'suse', ''
+] + RHEL_DISTROS + BSD_DISTROS
+
+if not VALID_DISTROS.contains(DISTRO)
+  error(
+    'Invalid distro supplied: "@0@". Must be one of: @1@'.format(
+      DISTRO, VALID_DISTROS)
+  )
+endif
+
+
+if BSD_DISTROS.contains(system) or BSD_DISTROS.contains(DISTRO)
+  if INIT_SYSTEM = ''
+      if DISTRO == ''
+          DISTRO = system
+      endif
+      INIT_SYSTEM = 'sysvinit_' + DISTRO
+  endif
+else
+  if INIT_SYSTEM == ''
+    INIT_SYSTEM = 'systemd'
+  endif
+  if DISTRO == ''
+    fs = import('fs')
+    if fs.is_file('/etc/redhat-release')
+      message('Setting distro to rhel because of host /etc/redhat-release')
+      DISTRO = 'rhel'
+    endif
+  endif
+endif
+
+if RHEL_DISTROS.contains(DISTRO)
+  lib_exec_dir = join_paths('/usr', get_option('libexecdir'), 'cloud-init')
+else
+  lib_exec_dir = join_paths('/usr', 'lib', 'cloud-init')
+endif
+message('LIBEXEC: @0@ @1@'.format(lib_exec_dir, DISTRO))
+
+
+pymod = import('python')
+python = pymod.find_installation('python3', required: true)
+python.dependency()
+
+find = find_program('find')
+
+# TODO(When RHEL support meson >= 1.0.0 use install_sources 'preserve_path: true')
+py_sources = run_command(find, 'cloudinit', '-maxdepth', '1', '-name', '*.py', check: true).stdout().strip().split('\n')
+py_sources_files = files(py_sources)
+python.install_sources(py_sources, subdir: 'cloudinit')
+
+py_source_dirs = run_command(find, 'cloudinit', '-type', 'd',  '-not', '-name', '*pycache*', check: true).stdout().strip().split('\n')
```

Comment:
Using meson as a whole scripting language to do installs is a smell that maybe this is doing a little bit too much heavy lifting. Rather than shelling out and parsing the contents and doing manual iteration, I think [install_subdir()](https://mesonbuild.com/Reference-manual_functions.html#install_subdir) could simplify this quite a bit. What do you think?

Reply:
`install_subdir` is generally intended for non-python package delivery because with that you'll have to also determine the appropriate python.get_install_dir() and it also doesn't allow us to exclude certain files with globbing unless we enumerate each known pycache directory. Generally meson wants static declarations of all file paths as seen in projects like https://sources.debian.org/src/numpy/1:2.2.4+ds-1/numpy/meson.build/?hl=305#L270 and https://sources.debian.org/src/dbus-python/1.4.0-1/meson.build/?hl=135#L113. But, since we don't want to update our meson.build file with every single new python module, we prefer https://mesonbuild.com/FAQ.html#but-i-really-want-to-use-wildcards

Reply:
No problem, glad it helped.